File: rkf451.py
# ルンゲ・クッタ=フェールベルグ法
#   刻み幅自動制御
#
#   連立1階微分方程式
#
# sikinote(https://slpr.sakura.ne/jp/qp) 2020/04/22によります
#

import numpy as np
import math
import matplotlib.pyplot as plt

def grk(n, x, y):
    #計算する関数
    # n:    方程式階数
    # x:    変数
    # y:    解
    f = np.zeros(n)
    ganma = 0.15

    f[0] = y[1]                         #dy/dx = v
    f[1] = -2 * ganma * y[1] - y[0]     #dv/dx = -2 * ganma * v - y
    return f


def dkf45(N, x, y, h, tol):
    # ルンゲ・クッタ=フェールベルグ法
    # 刻み幅自動制御
    # N:    方程式階数
    # x:    変数
    # y:    解
    # h:    刻み幅
    # tol:  精度
    #帰り値
    #info = -2  異常　計算範囲をオーバー
    #     = -1  異常　刻み値が小さくなりすぎ　tolの再検討が必要
    #     =  0  正常進行中
    #     =  1  x終了値に到達し完了

    #hのチェック
    hmin = 1e-14
    hmax = 1

    info = 0
    flag = 0

    if h > hmax:
        h = hmax
    if h < -hmax:
        h = -hmax    
    
    if abs(x - xe) <= h:
        h = xe - x

    if abs(x - xe) <= hmin:
        info = 1
        flag = 1
        #return x, y, h, info

    #精度値の計算
    Sy = 0
    for n in range(N):
        Sy += y[n] ** 2
    Sy = math.sqrt(Sy)
    if Sy >= 1:
        err = tol * Sy
    else:
        err = tol
    #print("x ={:13.10f} tol ={:13.10f}".format(x, err))

    #ルンゲクッタ法のブッチャーテーブル準備
    #5次6段
    S = 6    #段数
    b = np.zeros((2, S))
    b[0] = (25/216, 0, 1408/2565, 2197/4104, -1/5, 0)           #4次
    b[1] = (16/135, 0, 6656/12825, 28561/56430, -9/50, 2/55)    #5次
    c = (0, 1/4, 3/8, 12/13, 1, 1/2)
    #Rb = b[1] - b[0]
    Rb = (1/360, 0, -128/4275, -2197/75240, 1/50, 2/55)         # 検算用

    ty = np.zeros(N)
    fn = np.zeros(N)
    K = np.zeros((N, S))
    R = np.zeros(N)
    #y5 = np.zeros(N)    #.astype(np.float64)
    #Rcnt = 0

    #計算ループ
    while flag == 0:
        #係数Kの計算
        for s in range(S):
            tx = x + c[s] * h
            for n in range(N):
                if s == 0:
                    ty[n] = y[n]
                else:
                    ty[n] = y[n] + c[s] * K[n][s - 1]
            fn = grk(N, tx, ty)
            for n in range(N):
                K[n][s] = h * fn[n]
                
        #step 4
        #4次での解と5次での解の差を求める
        # R = |x(4次)-x(5次)| / hN
        R = 0
        r = np.zeros(N)
        for n in range(N):
            for s in range(S):
                r[n] += Rb[s] * K[n][s]
            R += r[n] ** 2
        R = abs(math.sqrt(R) / h / N)
        #Rcnt += 1
        #print('{:3d}回目 刻み幅:h ={:10.7f}  差:R ={:13.10f}'.format(Rcnt, h, R))

        #step 5
        #4次での解を計算
        #for n in range(N):
            #y5[n] = y[n]
        if R <= err:
            #print('R計算回数', Rcnt)
            x += h  
            for n in range(N):
                for s in range(S):
                    y[n] += b[0][s] * K[n][s]   #4次
                    #y5[n] += b[1][s] * K[n][s] #5次　比較用
                #print('y4[{:1d}] ={:20.16f}  y5[{:1d}] ={:20.16f} 差={:20.16f}'.format(n, y[n], n, y5[n], y[n] - y5[n]))
            flag = 1

        #step 6
        if R >= 1e-20:
            delta = (err / (2 * R)) ** (1 / 4)
        else:
            #Rがほぼ0の時
            delta = 4
        #print('delta =', delta)

        #step 7
        if delta <= 0.1:
            #現在の刻み幅hは大きすぎるので1/10にする
            h = 0.1 * h
        elif delta >= 4:
            #現在の刻み幅は小さすぎるので4倍する
            h = 4 * h
        else:
            #刻み幅を適正値に修正する
            h = delta * h

        #step 8
        #刻み幅がmaxを超えないように
        if h > hmax:
            h = hmax
        if h < -hmax:
            h = -hmax

        #step 9
        if abs(xe - x) <= abs(h):
            #hをx終端値に合わせる
            h = xe - x
            if abs(h) <= hmin:
                info = 1
                flag = 1

        #計算が範囲外になっている場合
        if h <= 0 and (xe - x) >= 0:
            info = -2
            flag = 1
        elif h > 0 and (xe - x) <= 0:
            info = -2
            flag = 1
        
    
    return x, y, h, info



# MAIN ------------------------------------------------------------

x = 0
xe = 20     #x終了値
#初期値
N = 2
y = np.zeros(N)
y[0] = 1
y[1] = -0.15

gx = []
gy = []

tol = 1e-8  #計算精度
h = xe - x  #刻み
i = 0
info = 0
step = 200

while info <= 0:
    x, y, h, info = dkf45(N, x, y, h, tol)
    i += 1
    if i % step == 0 or info == 1:     #最終データも表示
        print('{:8d} x:{:9.5f} y:{:13.9f} v:{:13.9f}   h:{:9.6f}'.format(i, x, y[0], y[1], h))
        gx.append(x)
        gy.append(y[0])
        
print('計算回数 {:8d}回'.format(i))    


#グラフ表示
ganma = 0.15
v0 = -0.15
y0 = 1
A = 1       #####
alfa = 0    #####

gx2 = []
gy2 = []
for i in range(0, 20000):
    x = float(i / 1000)
    y =  A * math.exp(-ganma * x) * math.cos(math.sqrt(1 - ganma ** 2) * x - alfa)
    gx2.append(x)
    gy2.append(y)

plt.plot(gx, gy, linestyle = 'None', marker = 'x', markeredgecolor = 'r')
plt.plot(gx2, gy2, color = 'g')
plt.xlabel('x')
plt.ylabel('y')
plt.show()


#  stop
#end program main

File: rkf451.f90
program main
    implicit none
    integer::i,N,info
    double precision::x,h,tol,xbound
    double precision,allocatable::y(:)
    external::grk

    N=1 ! Number of differential equations
    allocate(y(1:N))

    x=0d0; xbound=10d0
    y(1)=1d0 !initial condition

    h=xbound-x
    i=0; info=0; tol=1d-8;
    do while(info.le.0)
        call drkf45(grk,x,h,N,y,xbound,info,tol)
        write(10,'(3e25.10e3)')x,y(1),h
        i=i+1
    enddo
    write(6,*)"Number of referenced times -->",i

    stop
end program main

subroutine grk(N,x,y,f)
    implicit none
    integer,intent(in)::N
    double precision,intent(in)::x,y(1:N)
    double precision,intent(out)::f(1:N)

    ! Solve
    !  d y(1) / dt = y(1) * cos(t)

    f(1)=y(1)*cos(x)

    return
end subroutine grk

!===============================

subroutine drkf45(grk,x,h,N,y,xbound,info,tol)
    ! if h < hmin, propagate forcibly with warning.
    !
    !-----------------
    !info = -9  (maybe path the discontinue points)
    !     =  0  (Running now)
    !     =  1  (x reach xbound)
    !-----------------
    !  
    implicit none
    integer,intent(in)::N
    double precision,intent(in)::xbound,tol
    double precision,intent(inout)::x,h,y(1:N)
    integer,intent(inout)::info

    integer::i,j,FLAG,key
    double precision::R,delta,tx,Sy,err
    double precision,allocatable::ty(:),K(:,:),tf(:)
    double precision,parameter::hmin=1d-14,hmax=0.5d0
    integer,parameter::s=6
    double precision::a(1:s,1:s),b1(1:s),b2(1:s),c(1:s),Rc(1:s)
    external::grk

    c(1:6)=(/0d0, 0.25d0, 0.375d0, 0.9230769230769230769230769230769230769231d0, 1d0, 0.5d0/)
    a(1:6,1:6)=0d0
    a(1,1:6)=(/0d0, 0d0, 0d0, 0d0, 0d0, 0d0/)
    a(2,1:6)=(/0.25d0, 0d0, 0d0, 0d0, 0d0, 0d0/)
    a(3,1:6)=(/0.09375d0, 0.28125d0, 0d0, 0d0, 0d0, 0d0/)
    a(4,1:6)=(/0.8793809740555302685480200273099681383705d0, -3.277196176604460628129267182521620391443d0, 3.320892125625853436504324078288575329995d0, 0d0, 0d0, 0d0/)
    a(5,1:6)=(/2.032407407407407407407407407407407407407d0,-8d0, 7.173489278752436647173489278752436647173d0, -0.2058966861598440545808966861598440545809d0, 0d0, 0d0/)
    a(6,1:6)=(/-0.2962962962962962962962962962962962962963d0,2d0, -1.381676413255360623781676413255360623782d0, 0.4529727095516569200779727095516569200780d0,-0.275d0,0d0/)
    b2(1:6)=(/0.1185185185185185185185185185185185185185d0, 0.d0, 0.5189863547758284600389863547758284600390d0, 0.5061314903420166578061314903420166578061d0, -0.18d0, 0.03636363636363636363636363636363636363636d0/)
    b1(1:6)=(/0.1157407407407407407407407407407407407407d0, 0d0, 0.5489278752436647173489278752436647173489d0, 0.5353313840155945419103313840155945419103d0, -0.2d0, 0d0/)
    Rc(1:6)=(/0.002777777777777777777777777777777777777778d0,0d0, -0.02994152046783625730994152046783625730994d0, -0.02919989367357788410419989367357788410420d0, 0.02d0, 0.03636363636363636363636363636363636363636d0/)

    key=0
    allocate(ty(1:N),tf(1:N),K(1:s,1:N))
    ty=0d0; tf=0d0; K=0d0

    if(abs(h).ge.hmax)then
        h=sign(1d0,h)*hmax
    endif

    if(h.ge.abs(xbound-x))h=xbound-x

    FLAG=1
    if(abs(x-xbound).le.hmin)then
        info=1
        FLAG=0
    endif

    do while(FLAG.eq.1)
        tx=x
        do j=1,s
            tx=x+c(j)*h
            ty(1:N)=y(1:N)
            do i=1,j-1
                ty(1:N)=ty(1:N)+K(i,1:N)*a(j,i)
            enddo
            call grk(N,tx,ty,tf) 
            K(j,1:N)=h*tf(1:N)
        enddo

        !step 4
        R=0d0
        do i=1,N
            R=R+(Rc(1)*K(1,i)+Rc(3)*K(3,i)+Rc(4)*K(4,i)+Rc(5)*K(5,i)+Rc(6)*K(6,i))**2d0
        enddo
        R=abs(dsqrt(R)/h/dble(N))

        Sy=0d0
        do i=1,N
            Sy=Sy+(y(i)*y(i))
        enddo
        Sy=dsqrt(Sy)
        if(Sy.ge.1d0)then
            err=tol*Sy
        else
            err=tol
        endif

        !step 5
        if(R.le.err.or.key.eq.1)then
            x=x+h
            y(1:N)=y(1:N)+b1(1)*K(1,1:N)+b1(3)*K(3,1:N)+b1(4)*K(4,1:N)+b1(5)*K(5,1:N)
            FLAG=0
        endif

        !step 6
        !  Avoid zero deviding.
        if(R.ge.1d-20)then
            delta=(err/(2d0*R))**0.25d0
        else
            delta=4d0
        endif

        !step 7
        if(delta.le.0.1d0)then
            !function changes dramatically.
            h=0.1d0*h
        elseif(delta.ge.4d0)then
            !function changes loosely.
            h=4d0*h
        else
            !function changes moderately.
            h=delta*h
        endif

        !step 8
        if(abs(h).ge.hmax)then
            h=sign(1d0,h)*hmax
        elseif(abs(h).lt.hmin)then
            h=sign(1d0,h)*hmin
            key=1
        endif

        !step 9
        if(abs(xbound-x).le.abs(h))then
            h=xbound-x
            if(abs(h).le.hmin)then
                info=1
                FLAG=0
            endif
        end if

        if(h.le.0d0.and.xbound-x.ge.0d0)then
            info=1
            FLAG=0
        elseif(h.ge.0d0.and.xbound-x.le.0d0)then
            info=1
            FLAG=0
        endif
    enddo

    if(key.eq.1)then
        write(6,'(A,f10.5,A,f10.5)') "Strange point between ",x-h," and ",x
        info=-9
    endif

    deallocate(ty,tf,K)
    return
end subroutine drkf45